controller/misc/manager.py

```diff
@@ -72,3 +81,113 @@ def __updater_update_to_newest() -> None:
+        [
+            sql_alchemy_to_dict(button)
+            for button in customer_button.get_by_org_id(org_id)
+        ]
+    )
+
+
+def finalize_customer_buttons(
+    buttons: List[Dict[str, str]],
+    for_wrapped: bool = False,
+    deobfuscate_url: bool = False,
+) -> Dict[str, str]:
+    key = "createdBy" if for_wrapped else "created_by"
+    key_name = "createdByName" if for_wrapped else "created_by_name"
+    key_org_name = "organizationName" if for_wrapped else "organization_name"
+
+    user_ids = {str(e[key]) for e in buttons}  # set comprehension
+    name_lookup = {u_id: kratos.resolve_user_name_by_id(u_id) for u_id in user_ids}
+
+    for e in buttons:
+        e[key_name] = name_lookup[str(e[key])]
+        e[key_name] = (
+            (e[key_name]["first"] + " " + e[key_name]["last"])
+            if e[key_name]
+            else "Unknown"
+        )
+        # name comes from the join with organization
+        e[key_org_name] = e["name"]
+        del e["name"]
+    if deobfuscate_url:
+        for e in buttons:
+            convert_config_url_key_with_base64(e["config"], False)
+    return buttons
+
+
+def check_config_for_type(
+    type: enums.CustomerButtonType, config: Dict[str, Any], raise_me: bool = True
+) -> str:
+    if not config:
+        return __raise_or_return(raise_me, "Button must have a config")
+    if config.get("icon") is None:
+        return __raise_or_return(raise_me, "Button must have an icon")
+    t = config.get("tooltip")
+    if t is not None and len(t) < 10:
+        return __raise_or_return(
+            raise_me, "Button tooltip should be at least 10 characters long"
+        )
+    if type == enums.CustomerButtonType.DATA_MAPPER:
+        if (url := config.get("url")) is None:
+            return __raise_or_return(raise_me, "No url provided for data mapper button")
+        if "localhost:9060" in url:
+            # localhost dev version needs to be mapped to the "refinery" localhost url
+            # it's not meant to be in the same network but for development purposes the start script adds the dev network!
+            url = url.replace("localhost:9060", "external-data-mapper:80")
+        if "?" in url:
+            url += "&ping=true"
+        else:
+            url += "?ping=true"
+        try:
+
+            # Note that python requests dont set the options preflight + origin so they are tested without CORS
+            x = requests.post(url, json={"rows": [[]]}, timeout=2)
+            if not x.ok or "pong" not in x.text:
+                return __raise_or_return(
+                    raise_me, f"URL {url} answered with {x.status_code}"
+                )
+        except Exception:
+            import traceback
+
+            print(traceback.format_exc(), flush=True)
+            return __raise_or_return(raise_me, f"URL {url} is not reachable")
+
+        return  # returns None so "no error"
+    return __raise_or_return(raise_me, f"Unknown customer button type: {type}")
```

Comment:
unreachable return

Reply:
intended for us devs as known fallback => wont be changed